spring-integration-core/src/main/java/org/springframework/integration/handler/BarrierMessageHandler.java

```diff
@@ -0,0 +1,192 @@
+		Assert.notNull(outputProcessor, "'messageGroupProcessor' cannot be null");
+		this.correlationStrategy = correlationStrategy;
+		this.messageGroupProcessor =  outputProcessor;
+		this.timeout = timeout;
+	}
+
+	@Override
+	protected Object handleRequestMessage(Message<?> requestMessage) {
+		Object key = this.correlationStrategy.getCorrelationKey(requestMessage);
+		if (key == null) {
+			throw new MessagingException(requestMessage, "Correlation Strategy returned null");
+		}
+		Thread existing = this.inProcess.putIfAbsent(key, Thread.currentThread());
+		if (existing != null) {
+			throw new MessagingException(requestMessage, "Correlation key ("
+					+ key + ") is already in use by " + existing.getName());
+		}
+		SynchronousQueue<Message<?>> syncQueue = createOrObtainQueue(key);
+		try {
+			Message<?> release = syncQueue.poll(this.timeout, TimeUnit.MILLISECONDS);
+			if (release != null) {
+				return processRelease(requestMessage, key, release);
+			}
+		}
+		catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+			throw new MessagingException(requestMessage, "Interrupted while waiting for release", e);
+		}
+		finally {
+			this.inProcess.remove(key);
+			this.suspensions.remove(key);
+		}
+		return null;
+	}
+
+	private Object processRelease(Message<?> requestMessage, Object key, Message<?> release) {
+		this.suspensions.remove(key);
+		if (release.getPayload() instanceof Throwable) {
+			throw new MessagingException(requestMessage, "Releasing flow returned a throwable",
+					(Throwable) release.getPayload());
+		}
+		else {
+			return buildResult(key, requestMessage, release);
+		}
+	}
+
+	/**
+	 * Override to change the default mechanism by which the inbound and release messages
+	 * are returned as a result.
+	 * @param key The correlation key.
+	 * @param requestMessage the inbound message.
+	 * @param release the release message.
+	 * @return the result.
+	 */
+	protected Object buildResult(Object key, Message<?> requestMessage, Message<?> release) {
+		SimpleMessageGroup group = new SimpleMessageGroup(key);
+		group.add(requestMessage);
+		group.add(release);
+		return this.messageGroupProcessor.processMessageGroup(group);
+	}
+
+	private SynchronousQueue<Message<?>> createOrObtainQueue(Object key) {
+		SynchronousQueue<Message<?>> syncQueue = new SynchronousQueue<Message<?>>();
+		SynchronousQueue<Message<?>> existing = this.suspensions.putIfAbsent(key, syncQueue);
+		if (existing != null) {
+			syncQueue = existing;
+		}
+		return syncQueue;
+	}
+
+	@Override
+	public void triggerAction(Message<?> message) {
+		Object key = this.correlationStrategy.getCorrelationKey(message);
+		if (key == null) {
+			throw new MessagingException(message, "Correlation Strategy returned null");
+		}
+		SynchronousQueue<Message<?>> syncQueue = createOrObtainQueue(key);
+		try {
+			if (syncQueue != null && !syncQueue.offer(message, timeout, TimeUnit.MILLISECONDS)) {
```

Comment:
To be honest I don't like this `blocking` solution.
The `trigger` Thread can't depend on the main Thread.
And in my opinion the `triggerAction` for non-existing `key` must be rejected with the `MessageRejectedException`.

In addition: how about to consider to use `LockRegistry` for the `key` to suspend other `Thread` instead of fail fast on the `inProcess`?